tests/test_tuple_1d.cpp

```diff
@@ -0,0 +1,319 @@
+        CHECK(m.id(vertices[0], PrimitiveType::Edge) == 0);
+        CHECK(m.id(vertices[5], PrimitiveType::Edge) == 4);
+    }
+    SECTION("edges")
+    {
+        const std::vector<Tuple> edges = m.get_all(PrimitiveType::Edge);
+        REQUIRE(edges.size() == 5);
+        CHECK(m.id(edges[0], PrimitiveType::Edge) == 0);
+        CHECK(m.id(edges[1], PrimitiveType::Edge) == 1);
+        CHECK(m.id(edges[2], PrimitiveType::Edge) == 2);
+        CHECK(m.id(edges[3], PrimitiveType::Edge) == 3);
+        CHECK(m.id(edges[4], PrimitiveType::Edge) == 4);
+    }
+}
+
+TEST_CASE("1D_loop_lines", "[tuple_generation], [tuple_1d]")
+{
+    DEBUG_EdgeMesh m = loop_lines();
+
+    SECTION("vertices")
+    {
+        const std::vector<Tuple> vertices = m.get_all(PrimitiveType::Vertex);
+        REQUIRE(vertices.size() == 6);
+        CHECK(m.id(vertices[0], PrimitiveType::Vertex) == 0);
+        CHECK(m.id(vertices[1], PrimitiveType::Vertex) == 1);
+        CHECK(m.id(vertices[2], PrimitiveType::Vertex) == 2);
+        CHECK(m.id(vertices[3], PrimitiveType::Vertex) == 3);
+        CHECK(m.id(vertices[4], PrimitiveType::Vertex) == 4);
+        CHECK(m.id(vertices[5], PrimitiveType::Vertex) == 5);
+    }
+    SECTION("edges")
+    {
+        const std::vector<Tuple> edges = m.get_all(PrimitiveType::Edge);
+        REQUIRE(edges.size() == 6);
+        CHECK(m.id(edges[0], PrimitiveType::Edge) == 0);
+        CHECK(m.id(edges[1], PrimitiveType::Edge) == 1);
+        CHECK(m.id(edges[2], PrimitiveType::Edge) == 2);
+        CHECK(m.id(edges[3], PrimitiveType::Edge) == 3);
+        CHECK(m.id(edges[4], PrimitiveType::Edge) == 4);
+        CHECK(m.id(edges[5], PrimitiveType::Edge) == 5);
+    }
+}
+
+
+TEST_CASE("1D_self_loop", "[tuple_generation], [tuple_1d]")
+{
+    DEBUG_EdgeMesh m = self_loop();
+
+    SECTION("vertices")
+    {
+        const std::vector<Tuple> vertices = m.get_all(PrimitiveType::Vertex);
+        REQUIRE(vertices.size() == 1);
+        CHECK(m._debug_id(vertices[0], PrimitiveType::Vertex) == 0);
+    }
+    SECTION("edges")
+    {
+        const std::vector<Tuple> edges = m.get_all(PrimitiveType::Edge);
+        REQUIRE(edges.size() == 1);
+    }
+}
+
+TEST_CASE("1D_random_switches", "[tuple_operation],[tuple_1d]")
+{
+    DEBUG_EdgeMesh m = loop_lines();
+
+    SECTION("vertices")
+    {
+        const std::vector<Tuple> vertex_tuples = m.get_all(PrimitiveType::Vertex);
+        for (size_t i = 0; i < vertex_tuples.size(); ++i) {
+            Tuple t = vertex_tuples[i];
+            for (size_t j = 0; j < 10; j++) {
+                switch (rand() % 2) {
+                case 0: t = m.switch_tuple(t, PrimitiveType::Vertex); break;
+                case 1:
+                    if (!m.is_boundary(t)) {
+                        t = m.switch_tuple(t, PrimitiveType::Edge);
+                    }
+                    break;
+                default: break;
+                }
+                CHECK(m.is_valid_slow(t));
```

Comment:
Also here it would be nice if you could just get the hash accessor once and reuse it. You can even define it outside of the sections.